sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -275,10 +272,22 @@ case class CachedRDDBuilder(
         storageLevel,
         cachedPlan.conf)
     }
-    val cached = cb.map { batch =>
-      sizeInBytesStats.add(batch.sizeInBytes)
-      rowCountStats.add(batch.numRows)
-      batch
+    val cached = cb.mapPartitionsInternal { it =>
+      new Iterator[CachedBatch] {
+        override def hasNext: Boolean = {
```

Comment:
what if the caller side invokes `hasNext` multiple times after it returns `false`?

Reply:
shall we use `TaskContext.addTaskCompletionListener`?

Reply:
sgtm